### PR TITLE
Adding RedirectToWwwRule and extension

### PR DESCRIPTION
**Pull request by clockworkmice**
Adding a simple way to enforce www. subdomain using a new redirect rule

Addresses #194

### CODE REVIEW

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+{
+    public class RedirectToWwwRule : IRule
+    {
+        public int StatusCode { get; set; }
+        public bool AllowLocalhost { get; set; } = true;
+
+        public virtual void ApplyRule(RewriteContext context)
+        {
+            var req = context.HttpContext.Request;
+            if (AllowLocalhost && req.Host.Host.StartsWith("localhost", StringComparison.OrdinalIgnoreCase))
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: I don't see why you need to do StartsWith here.

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                return;
+            }
+
+            if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
+            {
+                context.Result = RuleResult.ContinueRules;
+                return;
+            }
+
+            var newUrl = new StringBuilder().Append(req.Scheme).Append("://www.").Append(req.Host).Append(req.PathBase).Append(req.Path).Append(req.QueryString);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: Use UriHelpers.BuildAbsolute instead here. https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Extensions/UriHelper.cs#L48

File: src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs [hunk 3]
```diff
@@ -117,5 +117,36 @@ public static RewriteOptions AddRedirectToHttps(this RewriteOptions options, int
...
+            return AddRedirectToWww(options, statusCode: StatusCodes.Status301MovedPermanently);
+        }
+
+        /// <summary>
+        /// Redirect a request to www subdomain if the incoming request is non-www
+        /// </summary>
+        /// <param name="options">The <see cref="RewriteOptions"/>.</param>
+        public static RewriteOptions AddRedirectToWww(this RewriteOptions options)
+        {
+            return AddRedirectToWww(options, statusCode: StatusCodes.Status302Found);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: I think we are using a redirect status code of 307 from now on. @Tratcher ?

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs (hunk 1 above)
Comment (clockworkmice) [reply src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: As opposed to straight equals? I think you're right actually, I was using Value previously

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            }
+
+            if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
+            {
+                context.Result = RuleResult.ContinueRules;
+                return;
+            }
+
+            var wwwHost = new HostString($"www.{req.Host}");
+            var newUrl = UriHelper.BuildAbsolute(req.Scheme, wwwHost, req.PathBase, req.Path, req.QueryString);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: Add in the FragmentString here (req.FragmentString). It was a mistake for other rules to not include it.

File: src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs [hunk 5]
```diff
@@ -117,5 +117,36 @@ public static RewriteOptions AddRedirectToHttps(this RewriteOptions options, int
             options.Rules.Add(new RedirectToHttpsRule { StatusCode = statusCode, SSLPort = sslPort });
             return options;
         }
+
+        /// <summary>
+        /// Redirect a request to www subdomain if the incoming request is non-www, with returning a 301
+        /// status code for permanently redirected.
+        /// </summary>
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: I'd reword to 
``` 
Permanently redirects the request to the www subdomain if the request is is non-www.
```

File: src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs [hunk 6]
```diff
@@ -117,5 +117,36 @@ public static RewriteOptions AddRedirectToHttps(this RewriteOptions options, int
...
+        /// <param name="options">The <see cref="RewriteOptions"/>.</param>
+        /// <returns></returns>
+        public static RewriteOptions AddRedirectToWwwPermanent(this RewriteOptions options)
+        {
+            return AddRedirectToWww(options, statusCode: StatusCodes.Status301MovedPermanently);
+        }
+
+        /// <summary>
+        /// Redirect a request to www subdomain if the incoming request is non-www
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]:
```
Redirect the request to the www subdomain if the incoming request is non-www.
```

File: src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs [hunk 7]
```diff
@@ -117,5 +117,36 @@ public static RewriteOptions AddRedirectToHttps(this RewriteOptions options, int
...
+        /// </summary>
+        /// <param name="options">The <see cref="RewriteOptions"/>.</param>
+        /// <param name="statusCode">The status code to add to the response.</param>
+        public static RewriteOptions AddRedirectToWww(this RewriteOptions options, int statusCode)
+        {
+            options.Rules.Add(new RedirectToWwwRule { StatusCode = statusCode });
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: Is there a way to set AllowLocalhost? I think you need to add another extension:
```csharp
 public static RewriteOptions AddRedirectToWww(this RewriteOptions options, int statusCode, bool allowLocalhost)
```

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.AspNetCore.Rewrite.Logging;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Rewrite.Internal
+{
+    public class RedirectToWwwRule : IRule
+    {
+        public int StatusCode { get; set; }
+        public bool AllowLocalhost { get; set; } = true;
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: Bleh, I don't like how I designed this when I was an intern. Let's make StatusCode and AllowLocalhost private and set them in the constructor, like:
```csharp
public RedirectToWwwRule(int statusCode, bool allowLocalhost) ...
```

File: test/Microsoft.AspNetCore.Rewrite.Tests/MiddlewareTests.cs [hunk 9]
```diff
@@ -147,6 +147,93 @@ public class MiddlewareTests
...
+            {
+                app.UseRewriter(options);
+            });
+            var server = new TestServer(builder);
+
+            var response = await server.CreateClient().GetAsync(new Uri("https://localhost:443"));
+
+            Assert.Null(response.Headers.Location);
+        }
+
```
Comment (jkotalik) [inline test/Microsoft.AspNetCore.Rewrite.Tests/MiddlewareTests.cs]: Let's test a few more cases:
- Uri with http
- Uri with many parts, like "https://example.com/test?a=b#foobar" and verify www is added appropriately.
- Uri with just localhost (without port).

You can do this easily by making the test a Theory. See https://github.com/aspnet/BasicMiddleware/blob/dev/test/Microsoft.AspNetCore.HttpsPolicy.Tests/HstsMiddlewareTests.cs#L57

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs (hunk 8 above)
Comment (clockworkmice) [reply src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: That's I would have written it personally too. I was just trying to be consistent with RedirectToHttpsRule. Will change

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs (hunk 4 above)
Comment (clockworkmice) [reply src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: how would I get the # fragment at this point? We're on the server

File: src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs (hunk 7 above)
Comment (clockworkmice) [reply src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: If someone really wants to not make an exception for localhost, it can be set by adding the RedirectToWwwRule explicitly. I imagined it would be used so infrequently that I didn't include it in an extension method.

Having said that, a comment or even an optional param for the extension method would show the expected behaviour more though when someone is using it for the first time and wondering why it isn't working in IISExpress. What do you think?

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                return;
+            }
+
+            if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
+            {
+                context.Result = RuleResult.ContinueRules;
+                return;
+            }
+
+            var wwwHost = new HostString($"www.{req.Host}");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: Use Host.Value here. The encoding is different from Host.ToString.

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs (hunk 4 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: No, there's never a fragment on the request.

File: src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs [hunk 11]
```diff
@@ -117,5 +117,36 @@ public static RewriteOptions AddRedirectToHttps(this RewriteOptions options, int
...
+
+        /// <summary>
+        /// Redirect a request to www subdomain if the incoming request is non-www, with returning a 301
+        /// status code for permanently redirected.
+        /// </summary>
+        /// <param name="options">The <see cref="RewriteOptions"/>.</param>
+        /// <returns></returns>
+        public static RewriteOptions AddRedirectToWwwPermanent(this RewriteOptions options)
+        {
+            return AddRedirectToWww(options, statusCode: StatusCodes.Status301MovedPermanently);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: Let's use 308 for new code. It works better for POSTs.

File: src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs (hunk 7 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: An non-exposed option might as well not exist. Add an overload or remove the option.

Comment (clockworkmice) [reply src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: An optional param on all the extension methods or just this one?

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: An additional overload with all options. Optional params are discouraged in our public API.

Comment (clockworkmice) [reply src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: That's fair enough. Just seems like a lot of bloat for something that won't ever be set to false. Maybe keep the localhost logic but just not make it an option? What do you think? I'm happy with either

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Rewrite/RewriteOptionsExtensions.cs]: Sure, go ahead and remove the option.

**Comment by clockworkmice** [comment]
@Tratcher @jkotalik updated as per comments

File: src/Microsoft.AspNetCore.Rewrite/Extensions/RewriteMiddlewareLoggingExtensions.cs [hunk 12]
```diff
@@ -82,6 +83,11 @@ static RewriteMiddlewareLoggingExtensions()
                             LogLevel.Debug,
                             12,
                             "Request to {requestedUrl} was ended");
+
+            _redirectedToWww = LoggerMessage.Define(
+                            LogLevel.Information,
+                            13,
+                            "Request redirected to WWW");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Extensions/RewriteMiddlewareLoggingExtensions.cs]: lower case www

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+                return;
+            }
+
+            if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
+            {
+                context.Result = RuleResult.ContinueRules;
+                return;
+            }
+
+            var wwwHost = new HostString($"www.{req.Host.Value}");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: You'll get some pretty strange results if the host is 127.0.0.1:80 or [::1]:80. I don't think those are common enough to worry about though.

File: src/Microsoft.AspNetCore.Rewrite/Extensions/RewriteMiddlewareLoggingExtensions.cs (hunk 12 above)
Comment (clockworkmice) [reply src/Microsoft.AspNetCore.Rewrite/Extensions/RewriteMiddlewareLoggingExtensions.cs]: done

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs (hunk 13 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: We may want to consider adding a list of excluded hosts, like https://github.com/aspnet/BasicMiddleware/commit/892cebb27eb1487a10e07d70967e490ecb38bce8#diff-15f19c319158f42c2602a43b53630d87R43.

**Comment by jkotalik** [comment]
@clockworkmice thanks again for the contribution. This should be available by 2.1.0-preview2 or on our nightly builds shortly.

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs (hunk 13 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToWwwRule.cs]: No, if it came up you'd want to exclude IP addresses as a class of host, not exclude individual values.
